File: HappyDog.Infrastructure/Pagination.cs
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Text;

namespace HappyDog.Infrastructure
{
    public class Pagination<T>
    {
        public List<T> Data { get; set; }
        public int TotalItems { get; set; }
        public int TotalPages => (int)Math.Ceiling((decimal)TotalItems / Size);
        public int Size { get; set; } = 20;
        public int Page { get; set; }

        /// <summary>
        /// Previous text, default is «(&amp;laquo;)
        /// </summary>
        [JsonIgnore]
        public string PreviousText { get; set; } = "&laquo;";

        /// <summary>
        /// Previous text, default is »(&amp;raquo;)
        /// </summary>
        [JsonIgnore]
        public string NextText { get; set; } = "&raquo;";

        /// <summary>
        /// Display the page number of the first page, if the value is set.
        /// </summary>
        [JsonIgnore]
        public string FirstPageText { get; set; }

        /// <summary>
        /// Display the page number of the last page, if the value is set.
        /// </summary>
        [JsonIgnore]
        public string LastPageText { get; set; }

        /// <summary>
        /// In the last item of the page bar, Appen is displayed in the style of the page number.
        /// </summary>
        [JsonIgnore]
        public string Append => Page + " / " + TotalPages;

        /// <summary>
        /// Auto hide when TotalPages is less than 2
        /// </summary>
        [JsonIgnore]
        public bool AutoHide { get; set; } = true;

        public string GetPageNumberLinks(Func<int, string> pageUrl)
        {
            if (AutoHide && TotalPages == 1)
            {
                return string.Empty;
            }
            StringBuilder builder = new StringBuilder();
            builder.Append("<nav><ul class=\"pagination\">");
            if (Page > 1)
            {
                if (FirstPageText != null)
                {
                    builder
                        .Append("<li class=\"page-item\"><a class=\"page-link\" href=\"")
                        .Append(pageUrl(1))
                        .Append("\">")
                        .Append(FirstPageText)
                        .Append("</a></li>");
                }
                builder
                   .Append("<li class=\"page-item\"><a class=\"page-link\" href=\"")
                   .Append(pageUrl(Page - 1))
                   .Append("\">")
                   .Append(PreviousText)
                   .Append("</a></li>");
            }
            for (int i = 1; i <= TotalPages; i++)
            {
                if (Page == i)
                    builder
                        .Append("<li class=\"page-item active\"><span class=\"page-link\">")
                        .Append(i)
                        .Append("</span></li>");
                else
                    builder.Append("<li class=\"page-item\"><a class=\"page-link\" href=\"")
                        .Append(pageUrl(i))
                        .Append("\">")
                        .Append(i)
                        .Append("</a></li>");
            }
            if (Page < TotalPages)
            {
                builder
                    .Append("<li class=\"page-item\"><a class=\"page-link\" href=\"")
                    .Append(pageUrl(Page + 1))
                    .Append("\">")
                    .Append(NextText)
                    .Append("</a></li>");
                if (LastPageText != null)
                {
                    builder
                        .Append("<li class=\"page-item\"><a class=\"page-link\" href=\"")
                        .Append(pageUrl(TotalPages))
                        .Append("\">")
                        .Append(LastPageText)
                        .Append("</a></li>");
                }
            }
            if (Append != null)
            {
                builder
                    .Append("<li class=\"page-item\"><span class=\"page-link\">")
                    .Append(Append)
                    .Append("</span></li>");
            }
            builder.Append("</ul></nav>");
            return builder.ToString();
        }

        /// <summary>
        /// Get the elasticity pagination bar, the count of displayed page numbers will not exceed the specified value.
        /// </summary>
        /// <param name="pageUrl"></param>
        /// <param name="total">Total number of pages</param>
        /// <returns></returns>
        /// <exception cref="InvalidOperationException"></exception>
        public string GetFlexibleLinks(Func<int, string> pageUrl, int total)
        {
            if (AutoHide && TotalPages == 1)
            {
                return string.Empty;
            }
            if (Page > TotalPages || Page < 1)
            {
                return string.Empty;
            }
            if (total < 1)
            {
                throw new InvalidOperationException("Make sure that ElasticityCount is greater than 0");
            }
            StringBuilder builder = new StringBuilder();
            builder.Append("<nav><ul class=\"pagination\">");

            int left = (total - 1) / 2;
            int right = total - left - 1;
            int start, end;
            if (Page <= left)
            {
                start = 1;
                end = total > TotalPages ? TotalPages : total;
            }
            else if ((TotalPages - Page) < right)
            {
                end = TotalPages;
                start = end - total + 1;
                if (start < 1)
                    start = 1;
            }
            else
            {
                start = Page - left;
                end = Page + right;
            }

            if (Page > 1)
            {
                if (FirstPageText != null)
                {
                    builder
                        .Append("<li class=\"page-item\"><a class=\"page-link\" href=\"")
                        .Append(pageUrl(1))
                        .Append("\">")
                        .Append(FirstPageText)
                        .Append("</a></li>");
                }
                builder
                   .Append("<li class=\"page-item\"><a class=\"page-link\" href=\"")
                   .Append(pageUrl(Page - 1))
                   .Append("\">")
                   .Append(PreviousText)
                   .Append("</a></li>");
            }
            for (int i = start; i <= end; i++)
            {
                if (Page == i)
                    builder
                        .Append("<li class=\"page-item active\"><span class=\"page-link\">")
                        .Append(i)
                        .Append("</span></li>");
                else
                    builder
                        .Append("<li class=\"page-item\"><a class=\"page-link\" href=\"")
                        .Append(pageUrl(i))
                        .Append("\">")
                        .Append(i)
                        .Append("</a></li>");
            }

            if (Page < TotalPages)
            {
                builder
                    .Append("<li class=\"page-item\"><a class=\"page-link\" href=\"")
                    .Append(pageUrl(Page + 1))
                    .Append("\">")
                    .Append(NextText)
                    .Append("</a></li>");
                if (LastPageText != null)
                {
                    builder
                        .Append("<li class=\"page-item\"><a class=\"page-link\" href=\"")
                        .Append(pageUrl(TotalPages))
                        .Append("\">")
                        .Append(LastPageText)
                        .Append("</a></li>");
                }
            }
            if (Append != null)
            {
                builder
                    .Append("<li class=\"page-item\"><span class=\"page-link\">")
                    .Append(Append)
                    .Append("</span></li>");
            }
            builder.Append("</ul></nav>");
            return builder.ToString();
        }

        public string GetSimpleLinks(Func<int, string> pageUrl)
        {
            if (AutoHide && TotalPages == 1)
            {
                return string.Empty;
            }
            StringBuilder builder = new StringBuilder();
            builder.Append("<nav><ul class=\"pagination\">");
            if (Page <= 1)
            {
                builder
                    .Append("<li class=\"page-item disabled\"><span class=\"page-link\">")
                    .Append(PreviousText)
                    .Append("</span></li>");
            }
            else
            {
                int prev = Page - 1;
                builder
                    .Append("<li class=\"page-item\"><a class=\"page-link\" href=\"")
                    .Append(pageUrl(prev))
                    .Append("\">")
                    .Append(PreviousText)
                    .Append("</a></li>");
            }
            if (Page >= TotalPages)
            {
                builder
                    .Append("<li class=\"page-item disabled\"><span class=\"page-link\">")
                    .Append(NextText)
                    .Append("</span></li>");
            }
            else
            {
                int next = Page + 1;
                builder
                    .Append("<li class=\"page-item\"><a class=\"page-link\" href=\"")
                    .Append(pageUrl(next))
                    .Append("\">")
                    .Append(NextText)
                    .Append("</a></li>");
            }
            if (Append != null)
            {
                builder
                    .Append("<li class=\"page-item disabled\"><span class=\"page-link\">")
                    .Append(Append)
                    .Append("</span></li>");
            }
            builder.Append("</ul></nav>");
            return builder.ToString();
        }
    }
}

File: HappyDog.Java/src/main/java/wang/doghappy/java/util/Pagination.java
package wang.doghappy.java.util;

import org.apache.commons.lang3.builder.ToStringBuilder;

import java.util.List;
import java.util.function.Function;

public class Pagination<T> {
    public Pagination() {
        setPreviousText("&laquo;");
        setNextText("&raquo;");
        setSize(20);
    }

    private List<T> data;
    private int totalItems;
    private int size;
    private int page;
    private int totalPages;

    private String previousText;
    private String nextText;
    private String firstPageText;
    private String lastPageText;
    private boolean autoHide;

    public List<T> getData() {
        return data;
    }

    public void setData(List<T> data) {
        this.data = data;
    }

    public int getTotalItems() {
        return totalItems;
    }

    public void setTotalItems(int totalItems) {
        this.totalItems = totalItems;
    }

    public int getTotalPages() {
        totalPages = (int) Math.ceil((double) getTotalItems() / getSize());
        return totalPages;
    }

    public int getSize() {
        return size;
    }

    public void setSize(int size) {
        this.size = size;
    }

    public int getPage() {
        return page;
    }

    public void setPage(int page) {
        this.page = page;
    }

    public String getPreviousText() {
        return previousText;
    }

    public void setPreviousText(String previousText) {
        this.previousText = previousText;
    }

    public String getNextText() {
        return nextText;
    }

    public void setNextText(String nextText) {
        this.nextText = nextText;
    }

    public String getFirstPageText() {
        return firstPageText;
    }

    public void setFirstPageText(String firstPageText) {
        this.firstPageText = firstPageText;
    }

    public String getLastPageText() {
        return lastPageText;
    }

    public void setLastPageText(String lastPageText) {
        this.lastPageText = lastPageText;
    }

    public String getAppend() {
        return getPage() + " / " + getTotalPages();
    }

    public boolean isAutoHide() {
        return autoHide;
    }

    public void setAutoHide(boolean autoHide) {
        this.autoHide = autoHide;
    }

    public int getOffset() {
        return (getPage() - 1) * getSize();
    }

    public String getPageNumberLinks(Function<Integer, String> pageUrl) {
        if (isAutoHide() && getTotalPages() == 1) {
            return "";
        }
        StringBuilder builder = new StringBuilder();
        builder.append("<nav><ul class=\"pagination\">");
        if (getPage() > 1) {
            if (getFirstPageText() != null) {
                builder
                        .append("<li class=\"page-item\"><a class=\"page-link\" href=\"")
                        .append(pageUrl.apply(1))
                        .append("\">")
                        .append(getFirstPageText())
                        .append("</a></li>");
            }
            builder
                    .append("<li class=\"page-item\"><a class=\"page-link\" href=\"")
                    .append(pageUrl.apply(getPage() - 1))
                    .append("\">")
                    .append(getPreviousText())
                    .append("</a></li>");
        }
        for (int i = 1; i <= getTotalPages(); i++) {
            if (getPage() == i)
                builder
                        .append("<li class=\"page-item active\"><span class=\"page-link\">")
                        .append(i)
                        .append("</span></li>");
            else
                builder.append("<li class=\"page-item\"><a class=\"page-link\" href=\"")
                        .append(pageUrl.apply(i))
                        .append("\">")
                        .append(i)
                        .append("</a></li>");
        }
        if (getPage() < getTotalPages()) {
            builder
                    .append("<li class=\"page-item\"><a class=\"page-link\" href=\"")
                    .append(pageUrl.apply(getPage() + 1))
                    .append("\">")
                    .append(getNextText())
                    .append("</a></li>");
            if (getLastPageText() != null) {
                builder
                        .append("<li class=\"page-item\"><a class=\"page-link\" href=\"")
                        .append(pageUrl.apply(getTotalPages()))
                        .append("\">")
                        .append(getLastPageText())
                        .append("</a></li>");
            }
        }
        builder
                .append("<li class=\"page-item\"><span class=\"page-link\">")
                .append(getAppend())
                .append("</span></li>");
        builder.append("</ul></nav>");
        return builder.toString();
    }

    public String getFlexibleLinks(Function<Integer, String> pageUrl, int total) {
        if (isAutoHide() && getTotalPages() == 1) {
            return "";
        }
        if (getPage() > getTotalPages() || getPage() < 1) {
            return "";
        }
        if (total < 1) {
            throw new IllegalArgumentException("Make sure that ElasticityCount is greater than 0");
        }
        StringBuilder builder = new StringBuilder();
        builder.append("<nav><ul class=\"pagination\">");

        int left = (total - 1) / 2;
        int right = total - left - 1;
        int start, end;
        if (getPage() <= left) {
            start = 1;
            end = total > getTotalPages() ? getTotalPages() : total;
        } else if ((getTotalPages() - getPage()) < right) {
            end = getTotalPages();
            start = end - total + 1;
            if (start < 1)
                start = 1;
        } else {
            start = getPage() - left;
            end = getPage() + right;
        }

        if (getPage() > 1) {
            if (getFirstPageText() != null) {
                builder
                        .append("<li class=\"page-item\"><a class=\"page-link\" href=\"")
                        .append(pageUrl.apply(1))
                        .append("\">")
                        .append(getFirstPageText())
                        .append("</a></li>");
            }
            builder
                    .append("<li class=\"page-item\"><a class=\"page-link\" href=\"")
                    .append(pageUrl.apply(getPage() - 1))
                    .append("\">")
                    .append(getPreviousText())
                    .append("</a></li>");
        }
        for (int i = start; i <= end; i++) {
            if (getPage() == i)
                builder
                        .append("<li class=\"page-item active\"><span class=\"page-link\">")
                        .append(i)
                        .append("</span></li>");
            else
                builder
                        .append("<li class=\"page-item\"><a class=\"page-link\" href=\"")
                        .append(pageUrl.apply(i))
                        .append("\">")
                        .append(i)
                        .append("</a></li>");
        }

        if (getPage() < getTotalPages()) {
            builder
                    .append("<li class=\"page-item\"><a class=\"page-link\" href=\"")
                    .append(pageUrl.apply(getPage() + 1))
                    .append("\">")
                    .append(getNextText())
                    .append("</a></li>");
            if (getLastPageText() != null) {
                builder
                        .append("<li class=\"page-item\"><a class=\"page-link\" href=\"")
                        .append(pageUrl.apply(getTotalPages()))
                        .append("\">")
                        .append(getLastPageText())
                        .append("</a></li>");
            }
        }
        builder
                .append("<li class=\"page-item\"><span class=\"page-link\">")
                .append(getAppend())
                .append("</span></li>");
        builder.append("</ul></nav>");
        return builder.toString();
    }

    public String getSimpleLinks(Function<Integer, String> pageUrl) {
        if (isAutoHide() && getTotalPages() == 1) {
            return "";
        }
        StringBuilder builder = new StringBuilder();
        builder.append("<nav><ul class=\"pagination\">");
        if (getPage() <= 1) {
            builder
                    .append("<li class=\"page-item disabled\"><span class=\"page-link\">")
                    .append(getPreviousText())
                    .append("</span></li>");
        } else {
            int prev = getPage() - 1;
            builder
                    .append("<li class=\"page-item\"><a class=\"page-link\" href=\"")
                    .append(pageUrl.apply(prev))
                    .append("\">")
                    .append(getPreviousText())
                    .append("</a></li>");
        }
        if (getPage() >= getTotalPages()) {
            builder
                    .append("<li class=\"page-item disabled\"><span class=\"page-link\">")
                    .append(getNextText())
                    .append("</span></li>");
        } else {
            int next = getPage() + 1;
            builder
                    .append("<li class=\"page-item\"><a class=\"page-link\" href=\"")
                    .append(pageUrl.apply(next))
                    .append("\">")
                    .append(getNextText())
                    .append("</a></li>");
        }
        builder
                .append("<li class=\"page-item\"><span class=\"page-link\">")
                .append(getAppend())
                .append("</span></li>");
        builder.append("</ul></nav>");
        return builder.toString();
    }

    @Override
    public String toString() {
        return new ToStringBuilder(this)
                .append("data", data)
                .append("totalItems", totalItems)
                .append("size", size)
                .append("page", page)
                .append("previousText", previousText)
                .append("nextText", nextText)
                .append("firstPageText", firstPageText)
                .append("lastPageText", lastPageText)
                .append("autoHide", autoHide)
                .toString();
    }
}
